Improve introspection of memory allocations done without memory limits checks

azat opened the pull request:
- Introduce new profile events - `MemoryAllocatedWithoutCheck` and `MemoryAllocatedWithoutCheckBytes`
- Replace log with new trace event - `MemoryAllocatedWithoutCheck`. Logging may lead to deadlock and it is not good for analytics
- Add suppression for some existing code, that is OK for now (though maybe if this goes to a `trace_log`, we shouldn't suppress anything)
- And also **respect memory limits from some places that allocate >16MiB of RAM (radix sort, async inserts, aggregator, some functions and filelog - all these should be exception safe)**
- Also do not log significant allocations from C

### Changelog category (leave one):
- Improvement

### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):
Respect memory limits in some places that are known to allocate significant (>16MiB) amount of memory (sorting, async inserts, file log)

azat committed:
Improve introspection of memory allocations done without memory limits checks

First problem is that LOG_TEST is disabled for stateless tests

The second is that this code works only in debug (because we cannot log
from memory tracker), but we can collect at least ProfileEvents.

And let's also add a check into test runner (clickhouse-test) to report
anomalies.

antonio2368 reviewed (commented):
Maybe add new trace_log type and use TraceSender to send stacktrace where check was avoided?

serxa reviewed (approved):
This is great! And I think TraceSender addition would be helpful for finding the places, let's add it as well (here or in another PR)

azat commented:
Good idea!

And anyway apparently logging can lead to deadlocks (something similar to https://github.com/ClickHouse/ClickHouse/pull/86473 maybe), so we cannot use this, CI will became flaky, I will replace it with trace events

azat committed:
Sanitize places that can have big allocations w/ ignored MEMORY_LIMIT_EXCEEDED

- Add MemoryTrackerDebugBlockerInThread for some internal places that
  should not throw
- Add TrackedString implementation, a std::string that can throw, and
  use it in places where we can have a significant memory usage
- Use AllocatorWithMemoryTracking in RadixSort as well (can create big
  allocations)

I've found this from the log from CI [1].

  [1]: https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=87035&sha=cf9779523fb72328052b0f541fef6cbd1cf1cf45&name_0=PR&name_1=Stateless%20tests%20%28amd_debug%2C%20AsyncInsert%2C%20s3%20storage%2C%20parallel%29

azat committed:
Do not log significant allocations from C

azat committed:
Add information about allocations with blocked exceptions into trace_log

And also remove logging, since it may lead to server deadlocks now.

azat committed:
Use clang's ability to check does function argument names matches over tmp var

azat committed:
Fix tidy

azat commented:
Green CI, but, the tables did not created on ci-logs, since it is broken (cc @maxknv )

And due to green CI trace_log has not been captured (sigh)

azat committed:
Track allocations in h3kRing function

azat committed:
Do not sanitize big allocations from C

azat commented:
Ok, I found traces in stress jobs

Looks good, but I will add few more exceptions after analyzing traces

```
:) select trace_type, count() from `trace_log.tsv.zst` group by all

SELECT
    trace_type,
    count()
FROM `trace_log.tsv.zst`
GROUP BY ALL

Query id: 7a7bc883-a279-4438-a839-430ea218b102

   ┌─trace_type──────────────────┬─count()─┐
1. │ Real                        │ 1193537 │
2. │ CPU                         │    2718 │
3. │ Memory                      │   65831 │
4. │ MemorySample                │     165 │
5. │ MemoryPeak                  │   65752 │
6. │ MemoryAllocatedWithoutCheck │      28 │
   └─────────────────────────────┴─────────┘
```

azat committed:
Track allocations for places array in aggregator

This is to suppress the following MemoryAllocatedWithoutCheck traces:

    sym:       StackTrace::StackTrace()
    incrementAllocationWithoutCheck(long)
    MemoryTracker::allocImpl(long, bool, MemoryTracker*, double)
    MemoryTracker::allocImpl(long, bool, MemoryTracker*, double)
    CurrentMemoryTracker::allocImpl(long, bool)
    operator new[](unsigned long)
    void DB::Aggregator::executeImplBatch<false, DB::AggregationMethodStringNoCache<StringHashMap<char*, Allocator<true, true>>, false>, DB::ColumnsHashing::HashMethodString<PairNoInit<StringRef, char*>, char*, true, false, false, false>>(DB::AggregationMethodStringNoCache<StringHashMap<char*, Allocator<true, true>>, false>&, DB::ColumnsHashing::HashMethodString<PairNoInit<StringRef, char*>, char*, true, false, false, false>&, DB::Arena*, unsigned long, unsigned long, DB::Aggregator::AggregateFunctionInstruction*, bool, bool, bool, char*) const
    void DB::Aggregator::executeImpl<DB::AggregationMethodStringNoCache<StringHashMap<char*, Allocator<true, true>>, false>, DB::ColumnsHashing::HashMethodString<PairNoInit<StringRef, char*>, char*, true, false, false, false>>(DB::AggregationMethodStringNoCache<StringHashMap<char*, Allocator<true, true>>, false>&, DB::ColumnsHashing::HashMethodString<PairNoInit<StringRef, char*>, char*, true, false, false, false>&, DB::Arena*, unsigned long, unsigned long, DB::Aggregator::AggregateFunctionInstruction*, bool, bool, char*) const
    void DB::Aggregator::executeImpl<DB::AggregationMethodStringNoCache<StringHashMap<char*, Allocator<true, true>>, false>>(DB::AggregationMethodStringNoCache<StringHashMap<char*, Allocator<true, true>>, false>&, DB::Arena*, unsigned long, unsigned long, std::__1::vector<DB::IColumn const*, std::__1::allocator<DB::IColumn const*>>&, DB::Aggregator::AggregateFunctionInstruction*, DB::ColumnsHashing::LastElementCacheStats&, bool, bool, char*) const
    DB::Aggregator::executeImpl(DB::AggregatedDataVariants&, unsigned long, unsigned long, std::__1::vector<DB::IColumn const*, std::__1::allocator<DB::IColumn const*>>&, DB::Aggregator::AggregateFunctionInstruction*, bool, bool, char*) const
    DB::Aggregator::executeOnBlock(std::__1::vector<COW<DB::IColumn>::immutable_ptr<DB::IColumn>, std::__1::allocator<COW<DB::IColumn>::immutable_ptr<DB::IColumn>>>, unsigned long, unsigned long, DB::AggregatedDataVariants&, std::__1::vector<DB::IColumn const*, std::__1::allocator<DB::IColumn const*>>&, std::__1::vector<std::__1::vector<DB::IColumn const*, std::__1::allocator<DB::IColumn const*>>, std::__1::allocator<std::__1::vector<DB::IColumn const*, std::__1::allocator<DB::IColumn const*>>>>&, bool&) const
    DB::AggregatingTransform::consume(DB::Chunk)
    DB::AggregatingTransform::work()
    DB::ExecutionThreadContext::executeTask()
    DB::PipelineExecutor::executeStepImpl(unsigned long, DB::IAcquiredSlot*, std::__1::atomic<bool>*)
    DB::PipelineExecutor::executeSingleThread(unsigned long, DB::IAcquiredSlot*)
    DB::PipelineExecutor::executeImpl(unsigned long, bool)
    DB::PipelineExecutor::execute(unsigned long, bool)
    DB::CompletedPipelineExecutor::execute()
    DB::ProjectionDescription::calculate(DB::Block const&, std::__1::shared_ptr<DB::Context const>, DB::PODArray<unsigned long, 4096ul, Allocator<false, false>, 63ul, 64ul> const*) const
    DB::MergeTreeDataWriter::writeTempPartImpl(DB::BlockWithPartition&, std::__1::shared_ptr<DB::StorageInMemoryMetadata const>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, DB::SourcePartsSetForPatch, std::__1::shared_ptr<DB::Context const>, unsigned long)
    DB::MergeTreeDataWriter::writeTempPart(DB::BlockWithPartition&, std::__1::shared_ptr<DB::StorageInMemoryMetadata const>, std::__1::shared_ptr<DB::Context const>)
    DB::MergeTreeSink::consume(DB::Chunk&)
    DB::SinkToStorage::onConsume(DB::Chunk)
    void std::__1::__function::__policy_invoker<void ()>::__call_impl[abi:se190107]<std::__1::__function::__default_alloc_func<DB::ExceptionKeepingTransform::work()::$_1, void ()>>(std::__1::__function::__policy_storage const*)
    DB::runStep(std::__1::function<void ()>, std::__1::shared_ptr<DB::ThreadGroup>&)
    DB::ExceptionKeepingTransform::work()
    DB::ExecutionThreadContext::executeTask()
    DB::PipelineExecutor::executeStepImpl(unsigned long, DB::IAcquiredSlot*, std::__1::atomic<bool>*)
    DB::PipelineExecutor::executeSingleThread(unsigned long, DB::IAcquiredSlot*)
    DB::PipelineExecutor::executeImpl(unsigned long, bool)
    DB::PipelineExecutor::execute(unsigned long, bool)
    void std::__1::__function::__policy_invoker<void ()>::__call_impl[abi:se190107]<std::__1::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<true, true>::ThreadFromGlobalPoolImpl<DB::CompletedPipelineExecutor::execute()::$_0>(DB::CompletedPipelineExecutor::execute()::$_0&&)::'lambda'(), void ()>>(std::__1::__function::__policy_storage const*)
    ThreadPoolImpl<std::__1::thread>::ThreadFromThreadPool::worker()
    void* std::__1::__thread_proxy[abi:se190107]<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void (ThreadPoolImpl<std::__1::thread>::ThreadFromThreadPool::*)(), ThreadPoolImpl<std::__1::thread>::ThreadFromThreadPool*>>(void*)

    sum(size): 40084912 -- 40.08 million
    count():   1

azat commented on src/Common/ProfileEvents.cpp:
```diff
@@ -626,6 +626,8 @@ The server successfully detected this situation and will download merged part fr
     M(WriteBufferFromS3WaitInflightLimitMicroseconds, "Time spent on waiting while some of the current requests are done when its number reached the limit defined by s3_max_inflight_parts_for_one_file.", ValueType::Microseconds) \
     M(QueryMemoryLimitExceeded, "Number of times when memory limit exceeded for query.", ValueType::Number) \
+    M(MemoryAllocatedWithoutCheck, "Number of times memory has been allocated without checking for memory constraints.", ValueType::Number) \
+    M(MemoryAllocatedWithoutCheckBytes, "About of bytes has been allocated without checking for memory constraints.", ValueType::Number) \
```
:+1:

azat committed:
Use [[likely]]

azat committed:
Update src/Common/ProfileEvents.cpp

Co-authored-by: Antonio Andelic <antonio2368@users.noreply.github.com>

azat commented:
>[Build (fuzzers)](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=87035&sha=9622d9ca754084515c4574e652845f0f351cf0d5&name_0=PR&name_1=Build%20%28fuzzers%29)

Known issue

>[Unit tests (tsan)](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=87035&sha=9622d9ca754084515c4574e652845f0f351cf0d5&name_0=PR&name_1=Unit%20tests%20%28tsan%29)

- https://github.com/ClickHouse/ClickHouse/issues/87123

>[Stress test (amd_ubsan)](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=87035&sha=9622d9ca754084515c4574e652845f0f351cf0d5&name_0=PR&name_1=Stress%20test%20%28amd_ubsan%29)

- https://github.com/ClickHouse/ClickHouse/pull/84084
